File: es3/es3.go
package main

import (
	"fmt"
	"os"
)

type Agent struct {
	x int
	y int
}

func makeAgent(x, y int) Agent {
	return Agent{x, y}
}

type Map struct {
	xmin, ymin, xmax, ymax int
	width, height          int
	houses                 [][]int
}

func (m *Map) moveAgent(a *Agent, c byte) {
	switch c {
	case '^':
		a.y--
		if a.y < m.ymin {
			m.ymin = a.y
		}
	case 'v':
		a.y++
		if a.y > m.ymax {
			m.ymax = a.y
		}
	case '<':
		a.x--
		if a.x < m.xmin {
			m.xmin = a.x
		}
	case '>':
		a.x++
		if a.x > m.xmax {
			m.xmax = a.x
		}
	}
}

func (m *Map) visitHouse(a *Agent) {
	m.houses[a.y-m.ymin][a.x-m.xmin] += 1
}

func (m *Map) allocateHouses() {
	m.width = m.xmax - m.xmin + 1
	m.height = m.ymax - m.ymin + 1
	m.houses = make([][]int, m.height)
	for y := 0; y < m.height; y++ {
		m.houses[y] = make([]int, m.width)
	}
}

func (m *Map) countHouses(presents int) int {
	count := 0
	for y := 0; y < m.height; y++ {
		for x := 0; x < m.width; x++ {
			if m.houses[y][x] >= presents {
				count++
			}
		}
	}
	return count
}

func makeMap() Map {
	return Map{}
}

func main() {
	rv := mainWithInt()
	if rv != 0 {
		os.Exit(rv)
	}
}

func mainWithInt() int {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <filename>\n", os.Args[0])
		return -1
	}

	input, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot open %s for reading\n", os.Args[1])
		return -1
	}
	defer input.Close()

	m1 := makeMap()
	m2 := makeMap()
	santa1 := makeAgent(0, 0)
	santa2 := makeAgent(0, 0)
	robo2 := makeAgent(0, 0)

	for i := 0; ; i++ {
		var c byte
		_, err = fmt.Fscanf(input, "%c", &c)
		if err != nil {
			break
		}
		m1.moveAgent(&santa1, c)
		if i&1 == 0 {
			m2.moveAgent(&santa2, c)
		} else {
			m2.moveAgent(&robo2, c)
		}
	}

	m1.allocateHouses()
	m2.allocateHouses()

	santa1 = makeAgent(0, 0)
	m1.visitHouse(&santa1)

	santa2 = makeAgent(0, 0)
	robo2 = makeAgent(0, 0)
	m2.visitHouse(&santa2)
	m2.visitHouse(&robo2)

	input.Seek(0, 0)
	for i := 0; ; i++ {
		var c byte
		_, err = fmt.Fscanf(input, "%c", &c)
		if err != nil {
			break
		}
		m1.moveAgent(&santa1, c)
		m1.visitHouse(&santa1)
		if i&1 == 0 {
			m2.moveAgent(&santa2, c)
			m2.visitHouse(&santa2)
		} else {
			m2.moveAgent(&robo2, c)
			m2.visitHouse(&robo2)
		}
	}

	fmt.Println("Answer1:", m1.countHouses(1))
	fmt.Println("Answer2:", m2.countHouses(1))

	return 0
}

File: es3/es3.py
#!/usr/bin/env python3

import sys

class Agent(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def visitHouse(self, map):
        map.houses[self.y - map.ymin][self.x - map.xmin] += 1

class Map(object):
    def __init__(self):
        self.xmin = 0
        self.xmax = 0
        self.ymin = 0
        self.ymax = 0
        self.width = 0
        self.height = 0
        self.houses = None

    def moveAgent(self, agent, b):
        if b == '^':
            agent.y -= 1
            if agent.y < self.ymin:
                self.ymin = agent.y
        elif b == 'v':
            agent.y += 1
            if agent.y > self.ymax:
                self.ymax = agent.y
        elif b == '<':
            agent.x -= 1
            if agent.x < self.xmin:
                self.xmin = agent.x
        elif b == '>':
            agent.x += 1
            if agent.x > self.xmax:
                self.xmax = agent.x

    def createHouses(self):
        self.width = self.xmax - self.xmin + 1
        self.height = self.ymax - self.ymin + 1
        self.houses = [[0 for x in range(self.width)] for y in range(self.height)]

    def countHouses(self, presents):
        count = 0
        for y in range(self.height):
            for x in range(self.width):
                if self.houses[y][x] >= presents:
                    count += 1
        return count

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    line = ""
    try:
        with open(sys.argv[1], "rt") as f:
            line = f.read()
    except:
        print("Cannot open {} for reading".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    m1 = Map()
    m2 = Map()
    santa1 = Agent(0, 0)
    santa2 = Agent(0, 0)
    robo2 = Agent(0, 0)
    for i in range(len(line)):
        m1.moveAgent(santa1, line[i])
        if i & 1 == 0:
            m2.moveAgent(santa2, line[i])
        else:
            m2.moveAgent(robo2, line[i])

    m1.createHouses()
    santa1 = Agent(0, 0)
    santa1.visitHouse(m1)

    m2.createHouses()
    santa2 = Agent(0, 0)
    santa2.visitHouse(m2)
    robo2 = Agent(0, 0)
    robo2.visitHouse(m2)

    for i in range(len(line)):
        m1.moveAgent(santa1, line[i])
        santa1.visitHouse(m1)
        if i & 1 == 0:
            m2.moveAgent(santa2, line[i])
            santa2.visitHouse(m2)
        else:
            m2.moveAgent(robo2, line[i])
            robo2.visitHouse(m2)

    print("Answer1: {}".format(m1.countHouses(1)))
    print("Answer2: {}".format(m2.countHouses(1)))
